proposals/0000-integrations.md

```diff
@@ -0,0 +1,171 @@
+
+Adding a first-class integration system would unlock a few huge wins for Astro:
+1. Empower users to do more by extending Astro themselves (not blocked by what Astro core can/can't do).
+2. Empower users to do more by reusing shared extensions (easy to add "X" to an Astro project).
+3. Empower more user-land experimentation, reducing how much Astro core blocks what a user can/can't do with Astro.
+4. Organize our codebase by refactoring more logic into internal integrations or moved out of core entirely into external integrations.
+
+To illustrate this, compare Partytown's documentation for getting started wtih Astro vs. getting started with Nuxt:
+- Astro: [3 steps across frontmatter, template, and a custom npm script](https://partytown.builder.io/astro)
+- Nuxt: [1 step](https://partytown.builder.io/nuxt)
+
+Tailwind suffers from a similar difficult setup story in Astro.
+
+# Detailed design
+
+**Background:** This API was reached through weeks of experimentation of different designs (see alternatives below). To test the work, I designed and build the following integrations, which are useful for illustrating this RFC: 
+
+- **Renderers:** [`lit`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/lit/index.js), [`svelte`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/svelte/index.js), [`react`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/react/index.js), [`preact`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/preact/index.js), [`vue`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/vue/index.js), [`solid`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/solid/index.js)
+- **Libraries:** [`tailwind`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/tailwind/index.js), [`partytown`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/partytown/index.js), [`turbolinks`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/turbolinks/index.js)
+- **Features:** [`sitemap`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/sitemap/index.js)
+
+## Integration Usage API
+
+```js
+// astro.config.js
+export default ({
+	integrations: [
+		import('@astrojs/vue'),
+        // or, with options
+		[import('@astrojs/vue'), {/* options */}],
+        // or, as a static ESM import at the top of the file
+		vuePlugin({/* options */})
+	],
+});
+```
+
+## Integration Author API
+
+```ts
+export interface AstroIntegration {
+	name: string;
+	hooks: {
```

Comment:
Are there going to be non-`astro:` namespace hooks available here as well?

If not, I think the prefix is a bit redundant - and it could be safely assumed that anything defined in an `AstroIntegration` will be Astro related. In that case, it might be better to make the clarification/distinction between Vite hooks in the docs rather than in the API.

Reply:
> Are there going to be non-astro: namespace hooks available here as well?

I don't have an answer! This was me being defensive because I wasn't sure and wanted to err on the side of caution. I do agree that any hook we add support for is implicitly an "astro" hook. The two use-cases I could see this being useful for later were:
- future custom hooks ("storybook:build" or "segment:event", for integrations that extend other integrations).
- future support for vite hooks in an Astro integration ("vite:load", "vite:transform") as sugar for providing a Vite plugin in the `astro:config:setup` hook.)

I generally avoid worrying too much about future features, but in this case if the only cost is a few potentially redundant letters on an API that is mainly for advanced users, then I think that's a strong case for the more verbose style as written.